### PR TITLE
add environment variables to powershell profile

### PR DESCRIPTION
**Pull request by mwrock**
This adds all variables from `load-omnibus-toolchain` to the powershell `$profile` which allows one to build in a powershell environment and also avoids having to manually load the batch file

### CODE REVIEW

**Comment by schisamo** [comment]
Would it be easier to generate some sort of PowerShell script (in addition to the `.bat` file) that is sourced automatically?

**Comment by mwrock** [comment]
That's essentially what this does. Powershell has an automatic variable called `$profile` which is the path to the user's profile script. So this runs the original .bat file and dumps all the environment variables and then dynamically adds script to the `$profile` that generates the same variable assignments.

**Comment by yzl** [comment]
:+1:

**Comment by scotthain** [comment]
I agree with @yzl - it would be super cool if we could have the script hang out after it's been run, but I don't feel like it's a blocker for merging this PR...

**Comment by mwrock** [comment]
Are the above comments regarding this line?:
```
 "Remove-Item `$tempFile" >> $profile
```
This is the output of the environment variables after running the batch file. However, because the batch file is run to get this output, all the environment information is echo'd to the console when a user opens a powershell console. So it should be very clear what was exported.

**Comment by scotthain** [comment]
I think it's more about how we usually do a 2 step (drop file, invoke file) so someone logging into the box for a testing can see the contents of the script after the fact from the box itself. If I'm not mistaken, this runs the script in place and puts the results out to the console, but not the script itself... Is that correct @mwrock ?

**Comment by mwrock** [comment]
no not the script itself but that script is retained in the ,.bat file and stays on the box. The problem is if you run a .bat file in powershell, it creates a new cmd.exe process so the environmet variables get lost after running that .bat file from a ps console. This change runs that same script and then has that same cmd process pipe the output of `SET` to a temp file which is hust a dump of the current environment key/values. Then it parses that and sets the same values in the user's powershell session.

**Comment by scotthain** [comment]
Ahhhhh ok. That works for me!

**Comment by yzl** [comment]
@mwrock Your explanation makes total sense.   :+1:

**Comment by schisamo** [comment]
My only worry with all this is there is a lot of magic going on. I would prefer we have something like `load-omnibus-toolchain.ps1` that a user has to explicitly call from a PowerShell session to load the Omnibus build toolchain environment. In many cases I would not want the Omnibus build toolchain environment to be loaded with every PowerShell session (since it does things like muck with default Ruby and Git).

**Comment by mwrock** [comment]
I'm really sorry, but I think maybe I misunderstand the context in which this cookbook is used. My interpretation was that it is used ephemerally for building but based on a couple comments I have seen now it sounds like this is installed on long lived infra. Is that a correct assessment?

I would agree that loading the environment variables for each powershell session could be quite annoying and perhaps destructive if this is on boxes used for other purposes.

**Comment by schisamo** [comment]
@mwrock No need to be sorry, I'll try to fill you in on some required tribal knowledge! This cookbook is used in both scenarios you mentioned. We manage our CI build slaves with this cookbook *and* use it when converging development environments for omnibus projects. The hope is that both environments match since the same Chef code is used to manage them.

**Comment by mwrock** [comment]
I'm going to revise this PR so that it does not drop the environment into `$profile` but rather a `load-omnibus-toolchain.PS1`

**Comment by mwrock** [comment]
ok. updated this to create a `load-omnibus-toolchain.ps1` instead of setting up environment from the `$profile`

**Comment by yzl** [comment]
:+1:

File: recipes/_environment.rb [hunk 1]
```diff
@@ -107,6 +108,24 @@
     owner node['omnibus']['build_user']
     group node['omnibus']['build_user_group']
   end
+
+  powershell_file = batch_file.gsub('.bat', '.ps1')
```
Comment (schisamo) [inline recipes/_environment.rb]: Let's skip just keep this simple here:

```ruby
powershell_file = windows_safe_path_join(build_user_home, 'load-omnibus-toolchain.ps1')
```

File: recipes/_environment.rb [hunk 2]
```diff
@@ -107,6 +108,24 @@
...
+      cmd /c " `"#{batch_file}`" && set > `"$tempFile`" "
+      Get-Content $tempFile | Foreach-Object {
+        if($_ -match "^(.*?)=(.*)$")
+        {
+          Set-Content "env:\\$($matches[1])" $matches[2]
+        }
+      }
+
+      Remove-Item $tempFile
+    EOH
```
Comment (schisamo) [inline recipes/_environment.rb]: I'd prefer we skip any `*.bat` file loading magic and just set the environment variables directly in this PowerShell script like we do here:
https://github.com/chef-cookbooks/omnibus/blob/5a3b21582bfee5649f0026f0c2b971ef97d8e321/recipes/_environment.rb#L32-L35

**Comment by mwrock** [comment]
ok just made the changes to the ps script suggested by seth.

**Comment by schisamo** [comment]
This is great @mwrock! We just switched to managing this cookbook with Delivery. Can you please close this PR and open a changeset on delivery.chef.co? Ping me if you need help.

**Comment by schisamo** [comment]
@delivery review

**Comment by schisamo** [comment]
@delivery review

**Comment by schisamo** [comment]
@mwrock Would it make sense to resurrect this PR? Would be nice to have a `load-omnibus-toolchain.ps1` file next to our `load-omnibus-toolchain.bat` file.

**Comment by mwrock** [comment]
@schisamo I rebased and cleaned it up but need to test it. Can you reopen the PR? Thanks

**Comment by schisamo** [comment]
@mwrock I think you'll just have to open a new one as I the __Reopen__ button is greyed out and when I hover I see the message: _The branch was recreated or force-pushed recreated_.

**Comment by mwrock** [comment]
created PR #167 to revive this.

**Pull request closed without merging**